File: app/src/main/java/com/guerinet/mymartlet/ui/SplashActivity.java
/*
 * Copyright 2014-2017 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.guerinet.mymartlet.App;
import com.guerinet.mymartlet.BuildConfig;
import com.guerinet.mymartlet.R;
import com.guerinet.mymartlet.model.Transcript;
import com.guerinet.mymartlet.model.exception.MinervaException;
import com.guerinet.mymartlet.ui.dialog.DialogHelper;
import com.guerinet.mymartlet.ui.settings.AgreementActivity;
import com.guerinet.mymartlet.util.Constants;
import com.guerinet.mymartlet.util.dagger.prefs.PrefsModule;
import com.guerinet.mymartlet.util.dagger.prefs.UsernamePref;
import com.guerinet.mymartlet.util.dbflow.databases.StatementDB;
import com.guerinet.mymartlet.util.manager.HomepageManager;
import com.guerinet.mymartlet.util.manager.McGillManager;
import com.guerinet.mymartlet.util.manager.UpdateManager;
import com.guerinet.mymartlet.util.service.ConfigDownloadService;
import com.guerinet.mymartlet.util.storage.ClearManager;
import com.guerinet.mymartlet.util.thread.UserDownloader;
import com.guerinet.suitcase.prefs.BooleanPref;
import com.guerinet.suitcase.prefs.IntPref;
import com.guerinet.suitcase.util.Utils;
import com.orhanobut.hawk.Hawk;
import com.raizlabs.android.dbflow.sql.language.SQLite;

import java.io.IOException;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dagger.Lazy;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * First activity that is opened when the app is started
 * @author Julien Guerinet
 * @since 1.0.0
 */
public class SplashActivity extends BaseActivity {
    /**
     * Code used when starting the AgreementActivity
     */
    private static final int AGREEMENT_CODE = 100;
    /**
     * Container if the min version is not satisfied
     */
    @BindView(R.id.version_container)
    LinearLayout minVersionContainer;
    /**
     * Container with the login views
     */
    @BindView(R.id.login_container)
    LinearLayout loginContainer;
    /**
     * Container with the loading progress bar (when signing in or loading info for the first time)
     */
    @BindView(R.id.progress_container)
    LinearLayout progressContainer;
    /**
     * The login {@link Button}
     */
    @BindView(R.id.login_button)
    Button loginButton;
    /**
     * {@link EditText} where the user enters their username
     */
    @BindView(R.id.login_username)
    EditText usernameView;
    /**
     * {@link EditText} where the user enters their password
     */
    @BindView(R.id.login_password)
    EditText passwordView;
    /**
     * {@link CheckBox} where the user decides if their username should be remembered
     */
    @BindView(R.id.login_remember_username)
    CheckBox rememberUsername;
    /**
     * Update text for when downloading everything
     */
    @BindView(R.id.progress_text)
    TextView progressText;
    /**
     * The {@link McGillManager} instance
     */
    @Inject
    McGillManager mcGillManager;
    /**
     * {@link ClearManager} instance
     */
    @Inject
    ClearManager clearManager;
    /**
     * Remember username {@link BooleanPref}
     */
    @Inject
    @Named(PrefsModule.REMEMBER_USERNAME)
    BooleanPref rememberUsernamePref;
    /**
     * Min version {@link IntPref}
     */
    @Inject
    @Named(PrefsModule.MIN_VERSION)
    IntPref minVersionPref;
    /**
     * EULA {@link BooleanPref}
     */
    @Inject
    @Named(PrefsModule.EULA)
    BooleanPref eulaPref;

    @Inject
    UsernamePref usernamePref;
    /**
     * {@link UpdateManager} instance
     */
    @Inject
    UpdateManager updateManager;
    /**
     * The {@link HomepageManager} instance
     */
    @Inject
    HomepageManager homepageManager;
    /**
     * The {@link InputMethodManager}
     */
    @Inject
    Lazy<InputMethodManager> imm;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_splash);
        ButterKnife.bind(this);
        App.Companion.component(this).inject(this);

        new AppInitializer().execute();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case AGREEMENT_CODE:
                if (resultCode == RESULT_OK) {
                    // If they agreed, go to the next screen
                    showNextScreen();
                } else {
                    // If not, close the app
                    finish();
                }
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
                break;
        }
    }

    @Override
    protected void onReceivedBroadcast(Intent intent) {
        // Override the Minerva broadcast to not log the user out since they are already logged out
        if (!intent.getAction().equals(Constants.BROADCAST_MINERVA)) {
            super.onReceivedBroadcast(intent);
        }
    }

    /**
     * Shows the first screen to the user depending on their situation
     */
    private void showNextScreen() {
        if (minVersionPref.get() > BuildConfig.VERSION_CODE) {
            // If we don't have the min required version, show the right container
            minVersionContainer.setVisibility(View.VISIBLE);
        } else if (usernamePref.get() == null || !Hawk.contains(PrefsModule.Hawk.PASSWORD)) {
            // If we are missing some login info, show the login screen with no error message
            showLoginScreen((IOException) getIntent().getSerializableExtra(Constants.EXCEPTION));
        } else {
            // Try logging the user in and download their info
            new AutoLogin(true).execute();
        }
    }

    /**
     * Shows the login screen and an eventual error message
     *
     * @param e The error received when trying to login, null if none
     */
    private void showLoginScreen(@Nullable IOException e) {
        // Show the login container
        loginContainer.setVisibility(View.VISIBLE);

        // Delete of the previous user's info
        clearManager.all();

        // Fill out username text if it is present
        usernameView.setText(usernamePref.get());

        // Set it to that when clicking the IME Action button it tries to log you in directly
        passwordView.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_GO) {
                loginButton.performClick();
                return true;
            }
            return false;
        });

        // Remember Me box checked based on user's previous preference
        rememberUsername.setChecked(rememberUsernamePref.get());

        // Check if an error message needs to be displayed, display it if so
        if (e != null) {
            DialogHelper.error(this, (e instanceof MinervaException) ?
                    R.string.login_error_wrong_data : R.string.error_other);
        }

        analytics.sendScreen("Login");
    }

    /**
     * Called when the min version button is clicked
     */
    @OnClick(R.id.version_button)
    protected void downloadNewVersion() {
        // Redirect them to the Play Store
        Utils.openPlayStoreApp(this);
    }

    /**
     * Called when the login button is clicked
     */
    @OnClick(R.id.login_button)
    protected void login() {
        // Hide the keyboard
        usernameView.clearFocus();
        imm.get().hideSoftInputFromWindow(usernameView.getWindowToken(), 0);

        // Get the inputted username and password
        final String username = usernameView.getText().toString().trim();
        final String password = passwordView.getText().toString().trim();

        // Check that both of them are not empty, create appropriate error messages if so
        if (TextUtils.isEmpty(username)) {
            DialogHelper.error(this, R.string.login_error_username_empty);
            return;
        } else if (TextUtils.isEmpty(password)) {
            DialogHelper.error(this, R.string.login_error_password_empty);
            return;
        } else if (!Utils.isConnected(this)) {
            DialogHelper.error(this, R.string.error_no_internet);
            return;
        }

        progressContainer.setVisibility(View.VISIBLE);

        mcGillManager.login(username + getString(R.string.login_email), password,
                new Callback<ResponseBody>() {
                    @Override
                    public void onResponse(Call<ResponseBody> call,
                            Response<ResponseBody> response) {
                        // Store the login info
                        usernamePref.set(username);
                        Hawk.put(PrefsModule.Hawk.PASSWORD, password);
                        rememberUsernamePref.set(rememberUsername.isChecked());

                        analytics.sendEvent("Login", "Remember Username",
                                String.valueOf(rememberUsername.isChecked()));

                        handler.post(() -> {
                            // Hide the login container
                            loginContainer.setVisibility(View.GONE);

                            // Start the downloading of information
                            new AutoLogin(false).execute();
                        });
                    }

                    @Override
                    public void onFailure(Call<ResponseBody> call, Throwable t) {
                        int error = t instanceof MinervaException ?
                                R.string.login_error_wrong_data : R.string.error_other;

                        handler.post(() -> {
                            // If for some reason the activity is finishing, don't show this
                            if (isFinishing()) {
                                return;
                            }
                            progressContainer.setVisibility(View.GONE);
                            DialogHelper.error(SplashActivity.this, error);
                        });
                    }
                });
    }

    /**
     * Initializes the app by starting the config download and running through update code
     */
    private class AppInitializer extends AsyncTask<Void, Void, Void> {

        @Override
        protected Void doInBackground(Void... params) {
            // Run any update code
            updateManager.update();

            // Initialize the McGillService
            mcGillManager.init();

            // Start downloading the config
            startService(new Intent(SplashActivity.this, ConfigDownloadService.class));
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            if (!eulaPref.get()) {
                // If the user has not accepted the EULA, show it before continuing
                Intent intent = new Intent(SplashActivity.this, AgreementActivity.class)
                        .putExtra(PrefsModule.EULA, true);
                startActivityForResult(intent, AGREEMENT_CODE);
            } else {
                // If they have, go to the next screen
                showNextScreen();
            }
        }
    }

    /**
     * Initializes the app by logging the user in and downloading the required information
     */
    private class AutoLogin extends AsyncTask<Void, String, IOException> {
        /**
         * True if we should auto-login the user, false if they have just logged in
         */
        private boolean autoLogin;

        /**
         * Default Constructor
         *
         * @param autoLogin True if we should auto-login the user, false if they have just logged in
         */
        private AutoLogin(boolean autoLogin) {
            this.autoLogin = autoLogin;
        }

        @Override
        protected void onPreExecute() {
            // Show the progress container
            progressContainer.setVisibility(View.VISIBLE);

            // Reset the text (if it was set during a previous login attempt)
            progressText.setText("");
        }

        @Override
        protected IOException doInBackground(Void... params) {
            analytics.sendEvent("Splash", "Auto-Login", Boolean.toString(autoLogin));

            // If we're auto-logging in and there is no internet, skip everything
            if (autoLogin && !Utils.isConnected(SplashActivity.this)) {
                return null;
            }

            // Try logging them in if needed
            if (autoLogin) {
                try {
                    mcGillManager.login();
                } catch (IOException e) {
                    return e;
                }
            }

            // Check if we need to download everything or only the essential stuff
            //  We need to download everything if there is null info
            boolean downloadEverything =
                    SQLite.select().from(Transcript.class).querySingle() == null ||
                            !getDatabasePath(StatementDB.FULL_NAME).exists();

            // If we need to download everything, do it synchronously. If not, do it asynchronously
            UserDownloader userDownloader = new UserDownloader(SplashActivity.this) {
                @Override
                public void update(String section) {
                    publishProgress(section);
                }
            };

            if (downloadEverything) {
                try {
                    userDownloader.execute();
                } catch (IOException e) {
                    // If there was an exception, return it
                    return e;
                }
            } else {
                userDownloader.start();
            }
            return null;
        }

        @Override
        protected void onProgressUpdate(String... progress) {
            // Update the TextView
            progressText.setText(progress[0]);
        }

        @Override
        protected void onPostExecute(IOException e) {
            // Hide the container
            progressContainer.setVisibility(View.GONE);

            if (e == null) {
                // Connection successful: home page
                startActivity(new Intent(SplashActivity.this, homepageManager.getActivity()));
                finish();
            } else {
                // Connection not successful: login
                showLoginScreen(e);
            }
        }
    }
}

File: app/src/main/java/com/guerinet/mymartlet/ui/SplashActivity.kt
/*
 * Copyright 2014-2018 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import androidx.core.view.isVisible
import com.guerinet.mymartlet.BuildConfig
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.model.exception.MinervaException
import com.guerinet.mymartlet.model.transcript.Transcript
import com.guerinet.mymartlet.ui.settings.AgreementActivity
import com.guerinet.mymartlet.util.Constants
import com.guerinet.mymartlet.util.Prefs
import com.guerinet.mymartlet.util.extensions.errorDialog
import com.guerinet.mymartlet.util.manager.HomepageManager
import com.guerinet.mymartlet.util.manager.McGillManager
import com.guerinet.mymartlet.util.manager.UpdateManager
import com.guerinet.mymartlet.util.prefs.UsernamePref
import com.guerinet.mymartlet.util.retrofit.Result
import com.guerinet.mymartlet.util.service.ConfigDownloadService
import com.guerinet.mymartlet.util.thread.UserDownloader
import com.guerinet.suitcase.prefs.BooleanPref
import com.guerinet.suitcase.prefs.IntPref
import com.guerinet.suitcase.util.extensions.isConnected
import com.guerinet.suitcase.util.extensions.openPlayStoreApp
import com.orhanobut.hawk.Hawk
import com.raizlabs.android.dbflow.kotlinextensions.from
import com.raizlabs.android.dbflow.sql.language.SQLite
import kotlinx.android.synthetic.main.activity_splash.*
import org.jetbrains.anko.startActivityForResult
import org.jetbrains.anko.startService
import org.koin.android.ext.android.inject
import java.io.IOException

/**
 * First activity that is opened when the app is started
 * @author Julien Guerinet
 * @since 1.0.0
 */
class SplashActivity : BaseActivity() {

    private val mcGillManager: McGillManager by inject()

    private val rememberUsernamePref: BooleanPref by inject(Prefs.REMEMBER_USERNAME)

    private val minVersionPref: IntPref by inject(Prefs.MIN_VERSION)

    private val eulaPref: BooleanPref by inject(Prefs.EULA)

    private val usernamePref: UsernamePref by inject()

    private val imm: InputMethodManager by inject()

    private val updateManager: UpdateManager by inject()

    private val homePageManager: HomepageManager by inject()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_splash)

        // OnClick listeners
        versionButton.setOnClickListener { openPlayStoreApp() }
        loginButton.setOnClickListener {
            launch(UI) { loginPressed() }
        }

        launch {
            // Run any update code
            updateManager.update()

            // Initialize the McGillService
            mcGillManager.init()

            // Start downloading the Config
            startService<ConfigDownloadService>()

            if (!eulaPref.value) {
                // If the user has not accepted the EULA, show it before continuing
                startActivityForResult<AgreementActivity>(AGREEMENT_CODE, Prefs.EULA to true)
            } else {
                // If not, go to the next screen
                showNextScreen()
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        when (requestCode) {
            AGREEMENT_CODE -> if (resultCode == Activity.RESULT_OK) {
                // If they agreed, go to the next screen
                showNextScreen()
            } else {
                // If not, close the app
                finish()
            }
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun onReceivedBroadcast(intent: Intent) {
        // Override the Minerva broadcast to not log the user out since they are already logged out
        if (intent.action != Constants.BROADCAST_MINERVA) {
            super.onReceivedBroadcast(intent)
        }
    }

    /**
     * Shows the first screen to the user depending on their situation
     */
    private fun showNextScreen() {
        if (minVersionPref.value > BuildConfig.VERSION_CODE) {
            // If we don't have the min required version, show the right container
            minVersionContainer.isVisible = true
        } else if (usernamePref.value == null || !Hawk.contains(Prefs.PASSWORD)) {
            // If we are missing some login info, show the loginPressed screen with no error message
            showLoginScreen(intent.getSerializableExtra(Constants.EXCEPTION) as? IOException)
        } else {
            // Try logging the user in and download their info
            launch(UI) { login(true) }
        }
    }

    /**
     * Shows the login screen and an eventual error message [e]
     */
    private fun showLoginScreen(e: IOException?) {
        // Show the login container
        loginContainer.isVisible = true

        // Delete of the previous user's info
        clearManager.clearUserInfo()

        // Fill out username text if it is present
        username.setText(usernamePref.value)

        // Set it to that when clicking the IME Action button it tries to log you in directly
        password.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_GO) {
                loginButton.performClick()
                true
            } else {
                false
            }
        }

        // Remember Me box checked based on user's previous preference
        rememberUsername.isChecked = rememberUsernamePref.value

        // Check if an error message needs to be displayed, display it if so
        if (e != null) {
            errorDialog(if (e is MinervaException)
                R.string.login_error_wrong_data else R.string.error_other)
        }

        ga.sendScreen("Login")
    }

    /**
     * Called when the login button is pressed
     */
    private fun loginPressed() {
        // Hide the keyboard
        username.clearFocus()
        imm.hideSoftInputFromWindow(username.windowToken, 0)

        // Get the inputted username and password
        val username = this.username.text.toString().trim()
        val password = this.password.text.toString().trim()

        // Check that both of them are not empty, create appropriate error messages if so
        if (username.isEmpty()) {
            errorDialog(R.string.login_error_username_empty)
            return
        } else if (password.isEmpty()) {
            errorDialog(R.string.login_error_password_empty)
            return
        } else if (!isConnected) {
            errorDialog(R.string.error_no_internet)
            return
        }

        progressContainer.isVisible = true

        val result = mcGillManager.login(username + getString(R.string.login_email), password)

        when (result) {
            is Result.Success<*> -> {
                // Store the login info
                usernamePref.value = username
                Hawk.put(Prefs.PASSWORD, password)
                rememberUsernamePref.value = rememberUsername.isChecked

                ga.sendEvent("Login", "Remember Username", rememberUsername.isChecked.toString())

                // Hide the login container
                loginContainer.isVisible = false

                // Start the downloading of information
                login(false)
            }
            is Result.Failure -> {
                val error = if (result.exception is MinervaException)
                    R.string.login_error_wrong_data
                else
                    R.string.error_other

                // If for some reason the activity is finishing, don't show this
                if (isFinishing) {
                    return
                }

                progressContainer.isVisible = false
                errorDialog(error)
            }
        }
    }

    /**
     * Attempts to either [autoLogin] or manually log in the user
     */
    private fun login(autoLogin: Boolean) {
        // Show the progress container
        progressContainer.isVisible = true

        // Reset the progress text (if it was set during a previous login attempt
        progressText.text = ""

        ga.sendEvent("Splash", "Auto-Login", autoLogin.toString())

        // If we're auto-logging in and there's no internet, skip everything
        if (autoLogin && !isConnected) {
            openHomePage()
            return
        }

        // Try auto login if needed
        if (autoLogin) {
            val result = mcGillManager.login()
            if (result is Result.Failure) {
                // If auto login isn't successful, don't continue
                showLoginScreen(result.exception)
                return
            }
        }

        // Check if we need to download everything or only the essential stuff
        //  We need to download everything if there is null info
        val downloadEverything = SQLite.select().from(Transcript::class).querySingle() == null

        // If we need to download everything, do it synchronously. If not, do it asynchronously
        val userDownloader = object : UserDownloader(this@SplashActivity) {
            override fun update(section: String) {
                updateProgress(section)
            }
        }

        if (downloadEverything) {
            try {
                userDownloader.execute()
            } catch (e: IOException) {
                // If there was an exception, stop here
                showLoginScreen(e)
                return
            }
        } else {
            userDownloader.start()
        }

        // Hide the container
        progressContainer.isVisible = false

        // Connection successful: home page
        openHomePage()
    }

    /**
     * Updates the progress [message]
     */
    private fun updateProgress(message: String) {
        progressText.text = message
    }

    private fun openHomePage() {
        startActivity(Intent(this, homePageManager.activity))
        finish()
    }

    companion object {

        /**
         * Code used when starting the AgreementActivity
         */
        private const val AGREEMENT_CODE = 100
    }
}